### PR TITLE
Convert Lit-GPT LLaMA Model Format to Original LLaMA Format

### LINKED ISSUE #292: Convert Lit-GPT LLaMA Model Format to Original LLaMA Format
Conversion was handled in PR https://github.com/Lightning-AI/lit-gpt/pull/273; however tests need to be created for the conversion. This issue and its accompanying PR will handle the tests for converting Lit-GPT LLaMA format to its original format.

This is part of a series of issues and PRs to address #183.

### LINKED ISSUE #340: Saving my finetuned model
Hello, I was able to fine-tune Llama2 with the help of lit-gpt. Thanks! However, there seems to be **a mismatch between the keys in the state dictionary I am trying to load and the keys in the model's state dictionary**. My ultimate goal is to upload my model to Huggingface space. Upon running the code, I encountered the following error:

```
import transformers
from transformers import LlamaForCausalLM, LlamaTokenizer

# Load Tokenizer
tokenizer = LlamaTokenizer.from_pretrained('/content/lit-parrot/checkpoints/meta-llama/Llama-2-7b-chat-hf')

# Load Model
model = LlamaForCausalLM.from_pretrained('meta-llama/Llama-2-7b-chat-hf')
checkpoint = torch.load('/version_5/lit_model_adapter_finetuned.pth', map_location=torch.device('cpu'))
model_state_dict = checkpoint['model']
model.load_state_dict(model_state_dict)
torch.save(checkpoint['model'], '/checkpoints/model_weights.pth')

# model = LlamaForCausalLM.from_pretrained('/version_5/lit_model_adapter_finetuned.pth')
```

```
[/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py](https://localhost:8080/#) in load_state_dict(self, state_dict, strict, assign)
   2150 
   2151         if len(error_msgs) > 0:
-> 2152             raise RuntimeError('Error(s) in loading state_dict for {}:\n\t{}'.format(
   2153                                self.__class__.__name__, "\n\t".join(error_msgs)))
   2154         return _IncompatibleKeys(missing_keys, unexpected_keys)

RuntimeError: Error(s) in loading state_dict for LlamaForCausalLM:
	Missing key(s) in state_dict: "model.embed_tokens.weight", "model.layers.0.self_attn.q_proj.weight", "model.layers.0.self_attn.k_proj.weight", "model.layers.0.self_attn.v_proj.weight", "model.layers.0.self_attn.o_proj.weight", "model.layers.0.self_attn.rotary_emb.inv_freq", "model.layers.0.mlp.gate_proj.weight", "model.layers.0.mlp.up_proj.weight", "model.layers.0.mlp.down_proj.weight", "model.layers.0.input_layernorm.weight", "model.layers.0.post_attention_layernorm.weight", "model.layers.1.self_attn.q_proj.weight", "model.layers.1.self_attn.k_proj.weight", "model.layers.1.self_attn.v_proj.weight", "model.layers.1.self_attn.o_proj.weight", "model.layers.1.self_attn.rotary_emb.inv_freq", "model.layers.1.mlp.gate_proj.weight", "model.layers.1.mlp.up_proj.weight", "model.layers.1.mlp.down_proj.weight", "model.layers.1.input_layernorm.weight", "model.layers.1.post_attention_layernorm.weight", "model.layers.2.self_attn.q_proj.weight", "model.layers.2.self_attn.k_proj.weight", "model.layers.2.self_attn.v_proj.weight", "model.layers.2.self_attn.o_proj.weight", "model.layers.2.self_attn.rotary_emb.inv_freq", "model.layers.2.mlp.gate_proj.weight", "model.layers.2.mlp.up_proj.weight", "model.layers.2.mlp.down_proj.weight", "model.layers.2.input_layernorm.weight", "model.layers.2.post_attention_layernorm.weight", "model.layers.3.self_attn.q_proj.weight", "model.layers.3.self_attn.k_proj.weight", "model.layers.3.self_attn.v_proj.weight", "model.layers.3.self_attn.o_proj.weight", "mo...
	Unexpected key(s) in state_dict: "transformer.wte.weight", "transformer.h.0.norm_1.weight", "transformer.h.0.attn.attn.weight", "transformer.h.0.attn.proj.weight", "transformer.h.0.norm_2.weight", "transformer.h.0.mlp.fc_1.weight", "transformer.h.0.mlp.fc_2.weight", "transformer.h.0.mlp.proj.weight", "transformer.h.1.norm_1.weight", "transformer.h.1.attn.attn.weight", "transformer.h.1.attn.proj.weight", "transformer.h.1.norm_2.weight", "transformer.h.1.mlp.fc_1.weight", "transformer.h.1.mlp.fc_2.weight", "transformer.h.1.mlp.proj.weight", "transformer.h.2.norm_1.weight", "transformer.h.2.attn.gating_factor", "transformer.h.2.attn.attn.weight", "transformer.h.2.attn.proj.weight", "transformer.h.2.attn.adapter_wte.weight", "transformer.h.2.norm_2.weight", "transformer.h.2.mlp.fc_1.weight", "transformer.h.2.mlp.fc_2.weight", "transformer.h.2.mlp.proj.weight", "transformer.h.3.norm_1.weight", "transformer.h.3.attn.gating_factor", "transformer.h.3.attn.attn.weight", "transformer.h.3.attn.proj.weight", "transformer.h.3.attn.adapter_wte.weight", "transformer.h.3.norm_2.weight", "transformer.h.3.mlp.fc_1.weight", "transformer.h.3.mlp.fc_2.weight", "transformer.h.3.mlp.proj.weight", "transformer.h.4.norm_1.weight", "transformer.h.4.attn.gating_factor", "transformer.h.4.attn.attn.weight", "transformer.h.4.attn.proj.weight", "transformer.h.4.attn.adapter_wte.weight", "transformer.h.4.norm_2.weight", "transformer.h.4.mlp.fc_1.weight", "transformer.h.4.mlp.fc_2.weight", "transformer.h...
```

### PR DESCRIPTION
**Pull request by JustinGoheen**
This PR closes #292 by adding support for converting Lit-GPT llama models back to their original format.

Closes #292
Closes https://github.com/Lightning-AI/lit-gpt/issues/183
Closes #340

### CODE REVIEW

File: scripts/convert_lit_checkpoint.py [hunk 1]
```diff
@@ -100,6 +100,102 @@ def copy_weights_gpt_neox(
...
+        "transformer.h.{}.mlp.fc_1.weight": "model.layers.{}.mlp.gate_proj.weight",
+        "transformer.h.{}.mlp.fc_2.weight": "model.layers.{}.mlp.up_proj.weight",
+        "transformer.h.{}.mlp.proj.weight": "model.layers.{}.mlp.down_proj.weight",
+        "transformer.ln_f.weight": "model.norm.weight",
+        "lm_head.weight": "lm_head.weight",
+    }
+
+    for name, param in lit_weights.items():
+        # handle name
+        if "transformer.h" in name and not name.endswith(".attn.attn.weight"):
```
Comment (carmocca) [inline scripts/convert_lit_checkpoint.py:123]: If you put the elif below first, you'd avoid

```suggestion
        elif "transformer.h" in name:
```

File: scripts/convert_lit_checkpoint.py [hunk 2]
```diff
@@ -100,6 +100,102 @@ def copy_weights_gpt_neox(
...
+        elif name.endswith(".attn.attn.weight"):
+            from_name, number = layer_template(name, 2)
+            q = "model.layers.{}.self_attn.q_proj.weight".format(number)
+            k = "model.layers.{}.self_attn.k_proj.weight".format(number)
+            v = "model.layers.{}.self_attn.v_proj.weight".format(number)
+        else:
+            to_name = weight_map[name]
+
+        # handle param
+        if name.endswith(".attn.attn.weight"):
```
Comment (carmocca) [inline scripts/convert_lit_checkpoint.py:138]: Can't you put this code block together with the one above?

File: scripts/convert_lit_checkpoint.py [hunk 3]
```diff
@@ -100,6 +100,102 @@ def copy_weights_gpt_neox(
...
+        else:
+            param = load_param(param)
+            if saver is not None:
+                param = saver.store_early(param)
+            state_dict[to_name] = param
+
+
+def tensor_split(param: Union[torch.Tensor, NotYetLoadedTensor], config: Config, model_name: str) -> torch.Tensor:
+    if model_name != "llama":
+        raise NotImplementedError(f"{model_name}")
```
Comment (carmocca) [inline scripts/convert_lit_checkpoint.py:154]: This only gets called for llama. I suggest not doing this sort of validation

File: tests/test_convert_lit_checkpoint.py [hunk 4]
```diff
@@ -94,3 +94,77 @@ def test_against_original_gpt_neox():
...
+
+    theirs_model = GPTNeoXForCausalLM(theirs_config)
+    # assign must be set to True for torch.testing.assert_close to pass
+    theirs_model.load_state_dict(theirs_state_dict, strict=False, assign=True)
+
+    # test end to end
+    x = torch.tensor([[9856, 23, 491, 1536, 304]], dtype=torch.int32)
+    ours_y = ours_model(x)
+    theirs_y = theirs_model(x)["logits"]
+    torch.testing.assert_close(ours_y, theirs_y)
```
Comment (carmocca) [inline tests/test_convert_lit_checkpoint.py:126]: What is different in this addition compared to the already existing assertions for gpt-neox?

Comment (JustinGoheen) [reply tests/test_convert_lit_checkpoint.py:126]: Nothing. I had accidentally altered the test for GPT-NeoX and this is a reversion to fix that.

**Comment by JustinGoheen** [comment]
I've updated in a recent push that seems to have not hit the PR yet. 

https://github.com/Lightning-AI/lit-gpt/commit/a8a8eca553b21ff987713b2d09eb40df14fbd1a6

File: scripts/convert_lit_checkpoint.py (hunk 3 above)
Comment (JustinGoheen) [reply scripts/convert_lit_checkpoint.py]: updated in [a8a8eca](https://github.com/Lightning-AI/lit-gpt/pull/315/commits/a8a8eca553b21ff987713b2d09eb40df14fbd1a6)

File: scripts/convert_lit_checkpoint.py (hunk 2 above)
Comment (JustinGoheen) [reply scripts/convert_lit_checkpoint.py]: updated in [a8a8eca](https://github.com/Lightning-AI/lit-gpt/pull/315/commits/a8a8eca553b21ff987713b2d09eb40df14fbd1a6)

File: scripts/convert_lit_checkpoint.py (hunk 1 above)
Comment (JustinGoheen) [reply scripts/convert_lit_checkpoint.py]: updated in [a8a8eca](https://github.com/Lightning-AI/lit-gpt/pull/315/commits/a8a8eca553b21ff987713b2d09eb40df14fbd1a6)

**Comment by JustinGoheen** [comment]
Aside from issue #292, this PR could close #183.

File: tests/test_convert_lit_checkpoint.py (hunk 4 above)
Comment (carmocca) [reply tests/test_convert_lit_checkpoint.py:126]: It's showing up as duplicated: 
image

Comment (JustinGoheen) [reply tests/test_convert_lit_checkpoint.py]: fixed in [c3144fd](https://github.com/Lightning-AI/lit-gpt/pull/315/commits/c3144fd206ab55f50730c1943c7d086b411733c0)

File: scripts/convert_lit_checkpoint.py [hunk 5]
```diff
@@ -100,6 +100,98 @@ def copy_weights_gpt_neox(
         state_dict[to_name] = param
 
 
+def copy_weights_llama(
```
Comment (awaelchli) [inline scripts/convert_lit_checkpoint.py:103]: The function is not used anywhere. We need to wire it up in the

```py
 elif config._mlp_class == "LLaMAMLP":
        raise NotImplementedError(f"Conversion for {model_name} is not yet supported")
```

branch below in `convert_lit_checkpoint`, right?

Comment (JustinGoheen) [reply scripts/convert_lit_checkpoint.py:103]: updated in [ef64e6c](https://github.com/Lightning-AI/lit-gpt/pull/315/commits/ef64e6c2bf47aa3b1c99d11ef2f3d230a5b506e3)

File: tests/test_convert_lit_checkpoint.py [hunk 6]
```diff
@@ -101,3 +101,47 @@ def test_against_original_gpt_neox():
     ours_y = ours_model(x)
     theirs_y = theirs_model(x)["logits"]
     torch.testing.assert_close(ours_y, theirs_y)
+
+
+@torch.inference_mode()
+@pytest.mark.parametrize("size", ("7b", "70b"))
+def test_against_llama2(size):
```
Comment (awaelchli) [inline tests/test_convert_lit_checkpoint.py:108]: For consistency with the other test names
```suggestion
def test_against_original_llama2(size):
```

**Review by awaelchli — APPROVED** [review]

**Comment by JustinGoheen** [comment]
do not merge, my merge conflict resolution wasn't correct

**Comment by JustinGoheen** [comment]
@awaelchli my resolution of merge conflicts was done incorrectly. requesting a new review of the pushes I made to undo the merge.

**Pull request merged**